### PR TITLE
DOCS/FAQ: added build instructions

### PR DESCRIPTION
**Pull request by hoopoepg**
- added instructions how to build UCX-based apps for dynamic and static modes

### CODE REVIEW

File: docs/source/faq.md [hunk 1]
```diff
@@ -115,6 +115,34 @@ It allows customization of the various parameters. An environment variable
 has precedence over the value defined in `ucx.conf`.
 The file can be created using `ucx_info -Cf`.
 
+#### Build user application with UCX
+
+UCX supports metainformation subsystem based on `pkg-config` tool. To add UCX support
+into application output from `pkg-config` tool may be used in Makefile:
```
Comment (brminich) [inline docs/source/faq.md:121]: i'd add `,` before `output`

File: docs/source/faq.md [hunk 2]
```diff
@@ -115,6 +115,34 @@ It allows customization of the various parameters. An environment variable
...
+```
+program: program.c
+        cc program.c $(pkg-config --cflags --libs ucx)
+```
+To build static application where UCX is used particular TL-specific modules
```
Comment (brminich) [inline docs/source/faq.md:126]: maybe something like:
`To build static application with UCX supporting just specific network transport modules (e.g. CMA and KNEM), the following command can be used:`

File: docs/source/faq.md [hunk 3]
```diff
@@ -115,6 +115,34 @@ It allows customization of the various parameters. An environment variable
...
+should be added to command line, for example to build application with support
+CMA and KNEM command line may be used:
+```
+program: program.c
+        cc -static program.c $(pkg-config --cflags --libs --static ucx-cma ucx-knem ucx)
+```
+Currently supported modules are: `ucx-cma`, `ucx-ib`, `ucx-knem`, `ucx-xpmem` and `ucx-rdmacm`.
+Transports `TCP` and `self` are built-in into UCT and don't need additional
+compilation actions.
```
Comment (brminich) [inline docs/source/faq.md:135]: should we also mention shm transports?

File: docs/source/faq.md [hunk 4]
```diff
@@ -115,6 +115,33 @@ It allows customization of the various parameters. An environment variable
...
+modules (e.g. CMA and KNEM), the following command can be used:
+```
+program: program.c
+        cc -static program.c $(pkg-config --cflags --libs --static ucx-cma ucx-knem ucx)
+```
+Currently supported modules are: `ucx-cma`, `ucx-ib`, `ucx-knem`, `ucx-xpmem` and `ucx-rdmacm`.
+Transports `TCP` and `self` are built-in into UCT and don't need additional
+compilation actions.
+
+> **IMPORTANT NOTE:** static module ucx-ib depends from static libs `libnl` and
```
Comment (brminich) [inline docs/source/faq.md:136]: `depends on`?

**Review by brminich — DISMISSED** [review]

File: docs/source/faq.md (hunk 4 above)
Comment (hoopoepg) [reply docs/source/faq.md]: updated

File: docs/source/faq.md [hunk 5]
```diff
@@ -115,6 +115,33 @@ It allows customization of the various parameters. An environment variable
 has precedence over the value defined in `ucx.conf`.
 The file can be created using `ucx_info -Cf`.
 
+#### Build user application with UCX
+
+UCX supports metainformation subsystem based on `pkg-config` tool. To add UCX support
+into application, output from `pkg-config` tool may be used in Makefile:
```
Comment (yosefe) [inline docs/source/faq.md:121]: UCX supports metainformation subsystem based on `pkg-config` tool. In order to build the
application with UCX development libraries.
For example, this is how pkg-config can be incorporated in a Makefile-based build:

File: docs/source/faq.md [hunk 6]
```diff
@@ -115,6 +115,33 @@ It allows customization of the various parameters. An environment variable
...
+```
+program: program.c
+        cc program.c $(pkg-config --cflags --libs ucx)
+```
+To build static application with UCX supporting just specific network transport
```
Comment (yosefe) [inline docs/source/faq.md:126]: When linking with static UCX libraries, need to list explicitly all required transport modules.
For example, in order to support only cma and knem transports, can use this method:

File: docs/source/faq.md [hunk 7]
```diff
@@ -115,6 +115,33 @@ It allows customization of the various parameters. An environment variable
...
+modules (e.g. CMA and KNEM), the following command can be used:
+```
+program: program.c
+        cc -static program.c $(pkg-config --cflags --libs --static ucx-cma ucx-knem ucx)
+```
+Currently supported modules are: `ucx-cma`, `ucx-ib`, `ucx-knem`, `ucx-xpmem` and `ucx-rdmacm`.
```
Comment (yosefe) [inline docs/source/faq.md:132]: Currently, the following transport modules can be used with pkg-config:
| Package name  |  Provided transport service |
|-----------|-----------|
| ucx-cma | Shared memory using [cross memory attach](https://lwn.net/Articles/405284/) |
| ucx-knem | ... |
| ... | ... |

File: docs/source/faq.md [hunk 8]
```diff
@@ -115,6 +115,33 @@ It allows customization of the various parameters. An environment variable
...
+Transports `TCP`, `SHM` and `self` are built-in into UCT and don't need additional
+compilation actions.
```
Comment (yosefe) [inline docs/source/faq.md:134]: TCP, basic shared memory, and self transports are built-in into UCT and don't need additional
compilation actions.

File: docs/source/faq.md [hunk 9]
```diff
@@ -115,6 +115,33 @@ It allows customization of the various parameters. An environment variable
...
+
+> **IMPORTANT NOTE:** static module ucx-ib depends on static libs `libnl` and
```
Comment (yosefe) [inline docs/source/faq.md:136]: may add this note to the above table

File: docs/source/faq.md (hunk 8 above)
Comment (hoopoepg) [reply docs/source/faq.md]: reworded

File: docs/source/faq.md (hunk 9 above)
Comment (hoopoepg) [reply docs/source/faq.md:146]: hmmm, first we notice ucx-ib module and after this add important note. I think this location is more "gracefully"

Comment (yosefe) [reply docs/source/faq.md:146]: ok, BUT seems the table is not rendered as part of the note:
image

Comment (hoopoepg) [reply docs/source/faq.md:146]: yes, but it seems this is markdown limitation

Comment (hoopoepg) [reply docs/source/faq.md:146]: may be remove quoting from note?

Comment (yosefe) [reply docs/source/faq.md:146]: yes, make it a sub-header

Comment (hoopoepg) [reply docs/source/faq.md]: > yes, make it a sub-header

done

File: docs/source/faq.md [hunk 10]
```diff
@@ -143,10 +143,10 @@ Currently, the following transport modules can be used with pkg-config:
...
 need additional compilation actions.
 
-> **IMPORTANT NOTE:** static module ucx-ib depends on static libs `libnl` and
-`numactl`. Unfortunately such libs are not provided by binary Linux
-distributions. To workaround this libraries may be built and installed
-from source code:
+##### IMPORTANT NOTE:
+Static module ucx-ib depends on static libs `libnl` and `numactl`.
+Unfortunately such libs are not provided by binary Linux distributions. To
+workaround this libraries may be built and installed from source code:
```
Comment (yosefe) [inline docs/source/faq.md:149]: last reword:
```
The package ucx-ib depends requires static libraries for `libnl` and `numactl`, as a dependency of rdma-core. Most Linux distributions do not provide these static libraries by default, so they need to be built and installed manually. They can be downloaded from the following locations:
```

can squash after

File: docs/source/faq.md [hunk 11]
```diff
@@ -144,9 +144,10 @@ TCP, basic shared memory, and self transports are built-in into UCT and don't
 need additional compilation actions.
 
 ##### IMPORTANT NOTE:
-Static module ucx-ib depends on static libs `libnl` and `numactl`.
-Unfortunately such libs are not provided by binary Linux distributions. To
-workaround this libraries may be built and installed from source code:
+The package ucx-ib depends requires static libraries for `libnl` and `numactl`,
```
Comment (yosefe) [inline docs/source/faq.md:147]: sorry, pls remove "depends"

**Review by yosefe — APPROVED** [review]

**Comment by yosefe** [comment]
@hoopoepg pls squash

File: docs/source/faq.md (hunk 9 above)
Comment (shamisp) [reply docs/source/faq.md]: what about pkg for AMD ROCM, Nvidia CUDA, ugni ? I would expect it should not be very hard to add pkg conf for this transports ?

Comment (hoopoepg) [reply docs/source/faq.md]: adding pkg-configs for these transport itself is not an issue, but static dependency libs could produce a lot of investigations. we will look at this for next releases

**Review by brminich — APPROVED** [review]

File: docs/source/faq.md (hunk 9 above)
Comment (shamisp) [reply docs/source/faq.md]: I would suggest to include those now and just indicate that static build is not supported for those libraries.

Comment (hoopoepg) [reply docs/source/faq.md]: pkg-config files for transports useful for static build only. in dynamic mode UCX loads it in runtime, no need to make any special tricks for runtime

Comment (shamisp) [reply docs/source/faq.md]: But the tool is can be used for dynamic as well ? I think what your are saying that today it is only used for static builds.

Comment (hoopoepg) [reply docs/source/faq.md]: dynamic link suppose to use pkg-config `ucx` or `ucx-uct`, no additional modules should be added

**Comment by shamisp** [comment]
@yosefe it would be good to run FAQ changes by @tonycurtis

File: docs/source/faq.md [hunk 12]
```diff
@@ -115,6 +115,44 @@ It allows customization of the various parameters. An environment variable
 has precedence over the value defined in `ucx.conf`.
 The file can be created using `ucx_info -Cf`.
 
+#### Build user application with UCX
+
+UCX supports metainformation subsystem based on pkg-config tool. In order to
```
Comment (tonycurtis) [inline docs/source/faq.md:120]: ... a metainformation subsystem based on the pkg-config tool

File: docs/source/faq.md [hunk 13]
```diff
@@ -115,6 +115,44 @@ It allows customization of the various parameters. An environment variable
...
+build the application with UCX development libraries.  For example, this is how
```
Comment (tonycurtis) [inline docs/source/faq.md:121]: incomplete sentence

File: docs/source/faq.md [hunk 14]
```diff
@@ -115,6 +115,44 @@ It allows customization of the various parameters. An environment variable
...
+pkg-config can be incorporated in a Makefile-based build:
+```
+program: program.c
+        cc program.c $(pkg-config --cflags --libs ucx)
```
Comment (tonycurtis) [inline docs/source/faq.md:125]: $(shell pkg-config ...) ?

File: docs/source/faq.md [hunk 15]
```diff
@@ -115,6 +115,44 @@ It allows customization of the various parameters. An environment variable
...
+```
+When linking with static UCX libraries, need to list explicitly all required
```
Comment (tonycurtis) [inline docs/source/faq.md:127]: the user needs to list all required transport modules explicitly

File: docs/source/faq.md [hunk 16]
```diff
@@ -115,6 +115,44 @@ It allows customization of the various parameters. An environment variable
...
+transport modules.  For example, in order to support only cma and knem
+transports, can use this method:
```
Comment (tonycurtis) [inline docs/source/faq.md:129]: the user can use

File: docs/source/faq.md [hunk 17]
```diff
@@ -115,6 +115,44 @@ It allows customization of the various parameters. An environment variable
...
+```
+program: program.c
+        cc -static program.c $(pkg-config --cflags --libs --static ucx-cma ucx-knem ucx)
```
Comment (tonycurtis) [inline docs/source/faq.md:132]: make shell again?  and typically you'd use $(CC) instead of hardwired cc

File: docs/source/faq.md [hunk 18]
```diff
@@ -115,6 +115,44 @@ It allows customization of the various parameters. An environment variable
...
+Currently, the following transport modules can be used with pkg-config:
+| Package name  |  Provided transport service |
+|-----------|-----------|
+| ucx-cma | Shared memory using [Linux Cross-Memory Attach](https://lwn.net/Articles/405284/) |
+| ucx-knem | Shared memory using [High-Performance Intra-Node MPI Communication](https://knem.gitlabpages.inria.fr) |
+| ucx-xpmem | Shared memory using [XPMEM](https://github.com/hjelmn/xpmem) |
+| ucx-ib | [Infiniband](https://developer.nvidia.com/networking) based network transport |
+| ucx-rdmacm | Connection manager based on [RDMACM](https://github.com/ofiwg/librdmacm) |
+
+TCP, basic shared memory, and self transports are built-in into UCT and don't
```
Comment (tonycurtis) [inline docs/source/faq.md:143]: built into

File: docs/source/faq.md (hunk 14 above)
Comment (hoopoepg) [reply docs/source/faq.md:125]: updated

File: docs/source/faq.md (hunk 15 above)
Comment (hoopoepg) [reply docs/source/faq.md:127]: updated

File: docs/source/faq.md (hunk 16 above)
Comment (hoopoepg) [reply docs/source/faq.md:129]: updated

File: docs/source/faq.md (hunk 17 above)
Comment (hoopoepg) [reply docs/source/faq.md:132]: yes, fixed

File: docs/source/faq.md (hunk 18 above)
Comment (hoopoepg) [reply docs/source/faq.md:143]: fixed

**Pull request merged**